lib/rust/profiler/src/lib.rs

```diff
@@ -431,21 +542,165 @@ pub enum Event {
         /// When the event occurred.
         timestamp: Timestamp,
     },
+    /// Metadata: wrapper with dependency-injected contents.
+    Metadata(Metadata<M>),
 }
 
+impl<M, L> Event<M, L> {
+    /// Produce a new event that may have a different metadata type, with metadata values
+    /// converted by the given function.
+    fn map_metadata<F, M1>(self, mut f: F) -> Event<M1, L>
+    where F: FnMut(M) -> M1 {
+        match self {
+            Event::Start(start) => Event::Start(start),
+            Event::StartPaused(start) => Event::StartPaused(start),
+            Event::End { id, timestamp } => Event::End { id, timestamp },
+            Event::Pause { id, timestamp } => Event::Pause { id, timestamp },
+            Event::Resume { id, timestamp } => Event::Resume { id, timestamp },
+            Event::Metadata(Metadata { timestamp, data }) =>
+                Event::Metadata(Metadata { timestamp, data: f(data) }),
+        }
+    }
+}
+
+
+
+// ================
+// === Metadata ===
+// ================
+
+/// Wrapper adding a timestamp to dependency-injected contents.
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub struct Metadata<M> {
+    /// When the event occurred.
+    pub timestamp: Timestamp,
+    /// The data.
+    pub data:      M,
+}
+
+impl<M> From<Metadata<M>> for Event<M, &'static str> {
+    fn from(metadata: Metadata<M>) -> Self {
+        Event::Metadata(metadata)
+    }
+}
+
+
+// === OpaqueMetadata ===
+
+/// Black-box metadata object, for use when metadata doesn't need to be created/interpreted.
+#[derive(Debug, Copy, Clone, serde::Serialize, serde::Deserialize)]
+pub enum OpaqueMetadata {
+    /// Anything.
+    #[serde(other)]
+    Unknown,
+}
+
+/// Event type with opaque metadata.
+pub type CoreEvent = Event<OpaqueMetadata, &'static str>;
+
+
+// === ExternalMetadata ===
+
+/// Metadata stored separately.
+#[derive(Debug, Copy, Clone)]
+struct ExternalMetadata {
+    type_id: u32,
+}
+
+
+
+// ======================
+// === MetadataLogger ===
+// ======================
+
+/// An object that supports writing a specific type of metadata to the profiling log.
+#[derive(Debug)]
+pub struct MetadataLogger<T> {
+    id:      u32,
+    entries: rc::Rc<log::Log<T>>,
+}
+
+impl<T: 'static + serde::Serialize> MetadataLogger<T> {
+    /// Create a MetadataLogger for logging a particular type.
+    ///
+    /// The name given here must match the name used for deserialization.
+    pub fn register(name: &'static str) -> Self {
```

Comment:
I wonder whether this could not just be called `new`. From an API user perspective, I think it would be clearer to see that this is the default way to instantiate a `MetadataLogger`.

Reply:
I wanted to convey that it's a side-effectful operation, but that's probably obvious enough. I'll change it.